module/epochs/epoch_lookup_test.go

```diff
@@ -1,65 +1,255 @@
+}
 
-	for _, tc := range testCases {
-		epoch, err := lookup.EpochForView(tc.view)
-		require.NoError(t, err)
-		require.Equal(t, tc.epoch, epoch)
+// TestEpochForViewWithFallback_Curr tests constructing and subsequently querying
+// EpochLookup with an initial state of a current and next epoch.
+func (suite *EpochLookupSuite) TestEpochForViewWithFallback_CurrNext() {
+	epochs := []epochRange{suite.currEpoch, suite.nextEpoch}
+	suite.CommitEpochs(epochs...)
+	suite.CreateAndStartEpochLookup()
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, epochs...)
+}
+
+// TestEpochForViewWithFallback_Curr tests constructing and subsequently querying
+// EpochLookup with an initial state of a previous, current, and next epoch.
+func (suite *EpochLookupSuite) TestEpochForViewWithFallback_CurrNextPrev() {
+	epochs := []epochRange{suite.prevEpoch, suite.currEpoch, suite.nextEpoch}
+	suite.CommitEpochs(epochs...)
+	suite.CreateAndStartEpochLookup()
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, epochs...)
+}
+
+// TestEpochForViewWithFallback_Curr tests constructing and subsequently querying
+// EpochLookup with an initial state of epoch fallback triggered.
+func (suite *EpochLookupSuite) TestEpochForViewWithFallback_EpochFallbackTriggered() {
+	epochs := []epochRange{suite.prevEpoch, suite.currEpoch, suite.nextEpoch}
+	suite.CommitEpochs(epochs...)
+	suite.CreateAndStartEpochLookup()
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, epochs...)
+}
+
+func (suite *EpochLookupSuite) TestProtocolEvents_EpochFallbackTriggered() {
+	// initially, only current epoch is committed
+	suite.CommitEpochs(suite.currEpoch)
+	suite.CreateAndStartEpochLookup()
+
+	// trigger epoch fallback
+	suite.epochFallbackTriggered = true
+	suite.lookup.EpochEmergencyFallbackTriggered()
+
+	// wait for the protocol event to be processed (async)
+	assert.Eventually(suite.T(), func() bool {
+		_, err := suite.lookup.EpochForViewWithFallback(suite.currEpoch.finalView + 1)
+		return err == nil
+	}, 5*time.Second, 50*time.Millisecond)
+
+	// validate queries are answered correctly
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, suite.currEpoch)
+
+	// should handle multiple deliveries of the protocol event
+	suite.lookup.EpochEmergencyFallbackTriggered()
+	suite.lookup.EpochEmergencyFallbackTriggered()
+	suite.lookup.EpochEmergencyFallbackTriggered()
+
+	// validate queries are answered correctly
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, suite.currEpoch)
+}
+
+func (suite *EpochLookupSuite) TestProtocolEvents_CommittedEpoch() {
+	// initially, only current epoch is committed
+	suite.CommitEpochs(suite.currEpoch)
+	suite.CreateAndStartEpochLookup()
+
+	// commit the next epoch, and emit a protocol event
+	firstBlockOfCommittedPhase := unittest.BlockHeaderFixture()
+	suite.state.On("AtBlockID", firstBlockOfCommittedPhase.ID()).Return(suite.snapshot)
+	suite.CommitEpochs(suite.nextEpoch)
```

Comment:
## ❗ 
This modifies the value of `suite.phase = flow.EpochPhaseCommitted`, which has no concurrency protection. Very complex to review ... sill not quite sure that there are no data races. Also makes the test rely heavily on internal implementation details of `EpochLookup`.

Reply:
Good point - added concurrency protection for these fields in https://github.com/onflow/flow-go/pull/2927/commits/e22499eaecc465a98c996b3657798b6d0c09096c